File: Lectures & Talks/Day 3/Examples/mpi_comm.c
#include <mpi.h>
#include <stdio.h>
#include <math.h>

int main(int argc, char *argv[]) {

  int numtasks, rank, secret=0, root=0;
  int tag, dest, src;

  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD,&numtasks);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);

  if (rank == root) {
    secret = 42;
    // send secret to all process, but one
    MPI_Request req;
    for (dest=0;dest<numtasks-1;dest++) {
      tag=dest;
      MPI_Isend(&secret,1,MPI_INT,dest,tag,
	      MPI_COMM_WORLD,&req);
    }
  } else if (rank != numtasks-1) {
    // receive secret
    MPI_Status info;
    MPI_Recv(&secret,1,MPI_INT,
      MPI_ANY_SOURCE,MPI_ANY_TAG,MPI_COMM_WORLD,&info);    
    printf("process %d received data: from %d \
      with tag %d\n",rank,info.MPI_SOURCE,info.MPI_TAG);
  }

  // wait for all processes
  MPI_Barrier(MPI_COMM_WORLD);

  printf("rank %d secret: %d\n",rank,secret);

  MPI_Finalize();
}

File: Lectures & Talks/Day 3/Examples/mpi_comm.py
#!/usr/bin/env python

import numpy as np
import mpi4py; mpi4py.rc.recv_mprobe = False
from mpi4py import MPI

root = 0

comm = MPI.COMM_WORLD
numtasks = comm.Get_size()
rank = comm.Get_rank()

secret = np.zeros(1, dtype='i')

if (rank == root):
  secret[0] = 42
  # send secret to all process, but one
  for dest in range(1,numtasks-1):
    req=comm.Isend([secret,1,MPI.INT],dest=dest,tag=dest)

elif (rank != numtasks-1):
  info = MPI.Status()
  # receive secret
  comm.Recv(secret,source=MPI.ANY_SOURCE,
            tag=MPI.ANY_TAG,status=info)
  print("process",rank,"received data:",
        "from",info.Get_source(),"with tag",info.Get_tag())

# wait for all processes
comm.Barrier()

print("process",rank,"secret:",secret)
